## Import the necessary libraries

In [1]:
# this will help in making the Python code more structured automatically (good coding practice)
%load_ext nb_black

# Libraries to help with reading and manipulating data
import numpy as np
import pandas as pd

# Libraries to help with data visualization
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

# Removes the limit from the number of displayed columns and rows.
# This is so I can see the entire dataframe when I print it
pd.set_option("display.max_columns", None)
# pd.set_option('display.max_rows', None)
pd.set_option("display.max_rows", 200)

import warnings

warnings.filterwarnings("ignore")

<IPython.core.display.Javascript object>

## Load and explore the Data

In [2]:
rawdata = pd.read_csv("movie_metadata.csv")

<IPython.core.display.Javascript object>

In [3]:
rawdata

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
5038  Color        Scott Smith                     1.0      87.0   
5039  Color                NaN                    43.0      43.0   
5040  Color   Benjamin Roberds                    13.0      76.0   
5041  Color        Daniel Hsia                    14.0     100.0   
5042  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
5038                      2.0                   318.0     Daphne Zuniga   
5039                      NaN                   319.0     Valorie Curry   
5040                      0.0                     0.0     Maxwell Moody   
5041                      0.0                   489.0     Daniel Henney   
5042                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
5038                   637.0          NaN                     Comedy|Drama   
5039                   841.0          NaN     Crime|Drama|Mystery|Thriller   
5040                     0.0          NaN            Drama|Horror|Thriller   
5041                   946.0      10443.0             Comedy|Drama|Romance   
5042                    86.0      85222.0                      Documentary   

         actor_1_name                                        movie_title  \
0         CCH Pounder                                            Avatar    
1         Johnny Depp          Pirates of the Caribbean: At World's End    
2     Christoph Waltz                                           Spectre    
3           Tom Hardy                             The Dark Knight Rises    
4         Doug Walker  Star Wars: Episode VII - The Force Awakens    ...   
...               ...                                                ...   
5038      Eric Mabius                           Signed Sealed Delivered    
5039      Natalie Zea                         The Following                
5040      Eva Boehnke                              A Plague So Pleasant    
5041        Alan Ruck                                  Shanghai Calling    
5042      John August                                 My Date with Drew    

      num_voted_users  cast_total_facebook_likes          actor_3_name  \
0              886204                       4834             Wes Studi   
1              471220                      48350        Jack Davenport   
2              275868                      11700      Stephanie Sigman   
3             1144337                     106759  Joseph Gordon-Levitt   
4                   8                        143                   N

<IPython.core.display.Javascript object>

Observations:
* 5043 observations with 28 attributes

In [4]:
rawdata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      5024 non-null   object 
 1   director_name              4939 non-null   object 
 2   num_critic_for_reviews     4993 non-null   float64
 3   duration                   5028 non-null   float64
 4   director_facebook_likes    4939 non-null   float64
 5   actor_3_facebook_likes     5020 non-null   float64
 6   actor_2_name               5030 non-null   object 
 7   actor_1_facebook_likes     5036 non-null   float64
 8   gross                      4159 non-null   float64
 9   genres                     5043 non-null   object 
 10  actor_1_name               5036 non-null   object 
 11  movie_title                5043 non-null   object 
 12  num_voted_users            5043 non-null   int64  
 13  cast_total_facebook_likes  5043 non-null   int64

<IPython.core.display.Javascript object>

In [5]:
df = rawdata.copy()

<IPython.core.display.Javascript object>

### Qestions

In [6]:
# 1.2. The duration column contains data on the film length. How many missing values are there in this column?
df["duration"].isnull().sum()

15

<IPython.core.display.Javascript object>

In [7]:
# df["duration"].value_counts(dropna=False)
# this is so I can see the different values, and catch any unvalid value like, negative durations, etc..
# this exploration shows the 15 NaN values and all other values in a descending order

<IPython.core.display.Javascript object>

Observations:
* There are 15 missing values in the duration column

In [8]:
# 1.3. Replace the missing values in the duration column with the median value for this column.
df["duration"].fillna(df["duration"].mean(), inplace=True)
print(f'The number of missing values is now {df["duration"].isnull().sum()}')

The number of missing values is now 0


<IPython.core.display.Javascript object>

In [9]:
# 1.4. What is the average film length?
# Give the answer as a floating-point figure rounded to two decimal places.
average_film_rounded = np.round(df["duration"].mean(), 2)
print(
    f"The average film lenght rounded to 2 decimals is {average_film_rounded} minutes"
)

The average film lenght rounded to 2 decimals is 107.2 minutes


<IPython.core.display.Javascript object>

1.5. Create a movie_duration_category column, which will contain three categories
depending on the film length:

* Category "1. <90" if the film is less than 90 minutes long
* Category "2. 90–120" if the film is between 90 minutes and two hours long (inclusively)
* Category "3. >120" if the film is more than two hours long

In [10]:
# can add custom labels
df["duration_bin"] = pd.cut(
    df["duration"], [-np.inf, 90, 120, np.inf], labels=["< 90", "90-120", ">120",],
)

df["duration_bin"].value_counts(1, dropna=False)

90-120    0.599445
>120      0.211580
< 90      0.188975
Name: duration_bin, dtype: float64

<IPython.core.display.Javascript object>

Obeservations:
* Only 18.9% of movies last less 90 minutes
* 21.11% of movies last more than 120 minutes
* 60% of movies last between 90-120 minutes

**1.6. Build a summary table for films released after 2000 (inclusively), to list the numbers of
films:**

* Table rows: year
* Table columns: movie duration category ("<90", "90–120", ">120")
* The year of release should be displayed in the YYYY format.


In [11]:
a = df[df["title_year"] >= 2000]
a = a[["title_year", "duration_bin"]]

a.title_year = a.title_year.astype(int)

# a["title_year"] = a["title_year"].astype(str).apply(lambda x: x.replace(".0", ""))
# alternative method to remove .0

a.set_index("title_year", inplace=True)
a.to_excel("table for films released after 2000 (inclusively).xlsx")

<IPython.core.display.Javascript object>

In [12]:
# 1.7. How many films between 90 minutes and two hours long were released in 2008?
films_90to120_2008 = df[df["title_year"] == 2008]
films_90to120_2008.head()

color      director_name  num_critic_for_reviews  duration  \
12  Color       Marc Forster                   403.0     106.0   
16  Color     Andrew Adamson                   258.0     150.0   
54  Color   Steven Spielberg                   470.0     122.0   
58  Color     Andrew Stanton                   421.0      98.0   
66  Color  Christopher Nolan                   645.0     152.0   

    director_facebook_likes  actor_3_facebook_likes          actor_2_name  \
12                    395.0                   393.0       Mathieu Amalric   
16                     80.0                   201.0  Pierfrancesco Favino   
54                  14000.0                  1000.0          Ray Winstone   
58                    475.0                   522.0          Fred Willard   
66                  22000.0                 11000.0          Heath Ledger   

    actor_1_facebook_likes        gross                             genres  \
12                   451.0  168368427.0                   Action|Adventure   
16                 22000.0  141614023.0    Action|Adventure|Family|Fantasy   
54                 11000.0  317011114.0           Action|Adventure|Fantasy   
58                  1000.0  223806889.0  Adventure|Animation|Family|Sci-Fi   
66                 23000.0  533316061.0        Action|Crime|Drama|Thriller   

          actor_1_name                                        movie_title  \
12  Giancarlo Giannini                                 Quantum of Solace    
16      Peter Dinklage          The Chronicles of Narnia: Prince Caspian    
54       Harrison Ford  Indiana Jones and the Kingdom of the Crystal S...   
58   John Ratzenberger                                            WALL·E    
66      Christian Bale                                   The Dark Knight    

    num_voted_users  cast_total_facebook_likes    actor_3_name  \
12           330784                       2023    Rory Kinnear   
16           149922                      22697  Damián Alcázar   
54           333847                      14959   Jim Broadbent   
58           718837                       2975     Jeff Garlin   
66          1676169                      57802  Morgan Freeman   

    facenumber_in_poster                                      plot_keywords  \
12                   1.0  action hero|attempted rape|bond girl|official ...   
16                   4.0  brother brother relationship|brother sister re...   
54                   2.0  cult figure|femme fatale|indiana jones|unsubti...   
58                   0.0                     earth|obesity|plant|robot|soil   
66                   0.0  based on comic book|dc comics|psychopath|star ...   

                                      movie_imdb_link  num_user_for_reviews  \
12  http://www.imdb.com/title/tt0830515/?ref_=fn_t...                1243.0   
16  http://www.imdb.com/title/tt0499448/?ref_=fn_t...                 438.0   
54  http://www.imdb.com/title/tt0367882/?ref_=fn_t...                2054.0   
58  http://www.imdb.com/title/tt0910970/?ref_=fn_t...                1043.0   
66  http://www.imdb.com/title/tt0468569/?ref_=fn_t...                4667.0   

   language country content_rating        budget  title_year  \
12  English      UK          PG-13  200000000.0$      2008.0   
16  English     USA             PG  225000000.0$      2008.0   
54  English     USA          PG-13  185000000.0$      2008.0   
58  English     USA              G  180000000.0$      2008.0   
66  English     USA          PG-13  185000000.0$      2008.0   

    actor_2_facebook_likes  imdb_score  aspect_ratio  movie_facebook_likes  \
12                   412.0         6.7          2.35                     0   
16                   216.0         6.6          2.35                     0   
54                  1000.0         6.2          2.35                  5000   
58                   729.0         8.4          2.35                 16000   
66                 13000.0         9.0          2.35                 37000   

   duration

<IPython.core.display.Javascript object>

In [13]:
films_90to120_2008["duration_bin"].value_counts()  # 1 for percentages)

90-120    147
< 90       42
>120       36
Name: duration_bin, dtype: int64

<IPython.core.display.Javascript object>

Obeservations:
* There where 147 films with duration between 90-120 mins released in 2008, 
* This represents 65% of the films released in that year

1.8. The plot_keywords column holds keywords characterizing the film's plot. Using the data
in this column, create a column called movie_plot_category, to contain four categories
depending on the key words in the column:

* Category "love_and_death" if the keywords include both "love" and "death"
* Category "love" if the keywords include the word "love"
* Category "death" if the keywords include the word "death"
* Category "other" if the keywords do not meet the conditions above


In [14]:
movie_plot_category = []
for i in df.plot_keywords:
    if "love" in str(i) and "death" in str(i):
        movie_plot_category.append("love_and_death")
    elif "love" in str(i):
        movie_plot_category.append("love")
    elif "death" in str(i):
        movie_plot_category.append("death")
    else:
        movie_plot_category.append("others")
df["movie_plot_category"] = movie_plot_category

<IPython.core.display.Javascript object>

### 1.9. The imdb_score column shows a viewer rating for the film. 
* Build a table to reflect the average rating of films depending on which movie_plot_category category they belong to.


In [15]:
b = pd.DataFrame(df.groupby(by="movie_plot_category")["imdb_score"].mean())
b

imdb_score
movie_plot_category            
death                  6.535465
love                   6.580769
love_and_death         6.506250
others                 6.431422

<IPython.core.display.Javascript object>

### 1.10. What is the average rating of films in the "love" category? 
*Give the answer as a floatingpoint figure rounded to two decimal places.

In [16]:
love_avg = np.round(float(b[b.index == "love"]["imdb_score"]), 2)
print(f"The average rating of films in the love category is {love_avg}")

The average rating of films in the love category is 6.58


<IPython.core.display.Javascript object>

In [17]:
# 1.11. The budget column contains the film's budget. What is the median budget for all the films listed?
# Give the answer as an integer.
df["budget"]

0       237000000.0$
1       300000000.0$
2       245000000.0$
3       250000000.0$
4               0.0$
            ...     
5038            0.0$
5039            0.0$
5040         1400.0$
5041            0.0$
5042         1100.0$
Name: budget, Length: 5043, dtype: object

<IPython.core.display.Javascript object>

In [18]:
def budget_to_num(budget_val):
    """This function takes in a string representing a budget and converts it to a number.
    If the input is already numeric, which probably means it's NaN,
    this function just returns np.nan."""

    if isinstance(budget_val, str):  # checks if `income_val` is a string
        if budget_val.endswith(".0$"):
            return int(budget_val.replace(".0$", ""))
    else:  # this happens when the current income is np.nan
        return np.nan


df["budget_value"] = df["budget"].apply(budget_to_num)
median = int(df["budget_value"].median())
print(f"The median budget for all the films listed is: $ {median}")

The median budget for all the films listed is: $ 15000000


<IPython.core.display.Javascript object>

**time for task 1: 30min**